README.md
# Web API Base SDK PHP

For use when creating either project specific SDKs or for use in our [Web API Base SDK Bundle](https://github.com/northamericanbancard/web-api-base-sdk-bundle).

## Getting Started

Follow the below instructions in order to successfully include and utilize this library within your project.

### Prerequisites

Ensure you are running PHP >= 5.5.

If any endpoints you are calling are to be secured via IAM, ensure you have the IAM user's 
access key and secret key. If required, you should also obtain the `x-api-key` header's API key (found in API Gateway).

If you are accessing api endpoints that are secured via JWT, then ensure you obtain the proper JWT for your project.

*_NOTE_*: Regardless if you are using AWS as your client, you will still have the ability to utilize the `x-api-key` header
to supply an API key, if you wish.

### Installing

#### Composer

1. Include this library in your project by running `composer require nab/web-api-base-sdk-php`.

#### API Calls:

The allowed calls you can make on the service are noted as:

1. httpGet(string $url, array $queryParams = [], array $headers = [], $body = null)
2. httpPost(string $url, array $queryParams = [], array $headers = [], $body = null)
2. httpPut(string $url, array $queryParams = [], array $headers = [], $body = null)

#### IAM Secured Endpoints

Example Calls:

```php
// SignatureV4 defaults to use-east-1, execute-api in the constructor.
$service = new AwsApiGatewayClient('http://example.com', new SignatureV4(), new Credentials('access_key', 'secret_key'), 'x-api-key'|null, [] /*optional guzzle config*/);
 
$service->httpGet('/some/path', ['foo' => 'bar', 'baz' => 'bing'], ['Accept' => 'application/json'])
$service->httpPost('/some/path', ['foo' => 'bar', 'baz' => 'bing'], ['Accept' => 'application/json'], '{"a": "b"}')
```

**_NOTE_** The client services are an extension of `GuzzleHttp\Client`, and therefore returns an instance of
`\Psr\Http\Message\ResponseInterface`. These classes do NOT throw exceptions on non-200 responses, but rather
suppress exceptions with the intent for you to retrieve error data from the status code and response body.

#### JWT Secured Endpoints

*_Note_*: JWT Tokens will exist in each request under the header `Authorization: Bearer` as `Authorization: Bearer <MY_TOKEN>`

Example Calls:

```php
$service = new JwtClient('http://example.com', 'my.jwt.token', 'x-api-key'|null, [] /*optional guzzle config*/);
 
$service->httpGet('/some/path', ['foo' => 'bar', 'baz' => 'bing'], ['Accept' => 'application/json'])
$service->httpPost('/some/path', ['foo' => 'bar', 'baz' => 'bing'], ['Accept' => 'application/json'], '{"a": "b"}')
```

#### Basic Auth Secured Endpoints

*_Note_*: Auth will exist in each request under the header `Authorization: Bearer` as `Authorization: Bearer <MY_HASH>`

Example Calls:

```php
$service = new JwtClient('http://example.com', 'username', 'password', 'x-api-key'|null, [] /*optional guzzle config*/);
 
$service->httpGet('/some/path', ['foo' => 'bar', 'baz' => 'bing'], ['Accept' => 'application/json'])
$service->httpPost('/some/path', ['foo' => 'bar', 'baz' => 'bing'], ['Accept' => 'application/json'], '{"a": "b"}')
```

#### Unsecured Endpoints

Example Calls:

```php
$service = new SimpleClient('http://example.com', 'x-api-key'|null, [] /*optional guzzle config*/);
 
$service->httpGet('/some/path', ['foo' => 'bar', 'baz' => 'bing'], ['Accept' => 'application/json'])
$service->httpPost('/some/path', ['foo' => 'bar', 'baz' => 'bing'], ['Accept' => 'application/json'], '{"a": "b"}')
```

**_NOTE_** The client services are an extension of `GuzzleHttp\Client`, and therefore returns an instance of
`\Psr\Http\Message\ResponseInterface`. These classes do NOT throw exceptions on non-200 responses, but rather
supress exceptions with the intent for you to retrieve error data from the status code and response body.

## Development

**_First Install/Make Introduction_**

1.  Ensure you have autoconf (v2.69 at min) installed.
2.  In the project root, run `make` to install your new project.
3.  View the Makefile, or Makefile.in for further commands and required inputs.
4.  Any edits to Makefile should be done in Makefile.in and then run `autoconf && ./configure && make`

## Testing and Maintenance

The following command should be run often, and should be run during every code review:

```bash
make test
```

OR

```bash
make docker-test
```

OR

```bash
make phpcs
```

## Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/northamericanbancard/web-api-base-sdk/tags). 

## License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.5.0] - 2018-05-09
### Added
- Added httpDelete method.

## [1.4.0] - 2018-05-08
### Added
- Opened up the ability to futz with Guzzle headers on instantiation of client - but does not set them on the base
instance.

tests/JwtClientTest.php
<?php

/**
 * Filename:    JwtClientTest.php
 * Created:     1/22/18, at 8:05 PM
 * @author      James Hollenbeck <jhollenbeck@nabancard.com>
 * @copyright   1992-2018 North American Bancard
 */

namespace NAB\Tests;

use NAB\Tests\ClientTestHelpers\JwtClientTestHelper;
use NAB\WebApiBaseSdk\ClientInterface;

class JwtClientTest extends AbstractClientTestBase
{
    public function clientCanHandleRequestDataProvider()
    {
        return [
            [
                'system_under_test' => new JwtClientTestHelper(
                    'example.com',
                    self::TOKEN,
                    'abc',
                    $this->getGuzzleConfig(ClientInterface::HTTP_METHOD_GET)
                ),
                'system_under_test_method' => 'httpGet',
                'query_data' => $this->GET_REQUEST_QUERY_PARAMS,
            ],
            [
                'system_under_test' => new JwtClientTestHelper(
                    'example.com',
                    self::TOKEN,
                    'abc',
                    $this->getGuzzleConfig(ClientInterface::HTTP_METHOD_POST)
                ),
                'system_under_test_method' => 'httpPost',
                'query_data' => [],
            ],
            [
                'system_under_test' => new JwtClientTestHelper(
                    'example.com',
                    self::TOKEN,
                    'abc',
                    $this->getGuzzleConfig(ClientInterface::HTTP_METHOD_PUT)
                ),
                'system_under_test_method' => 'httpPut',
                'query_data' => [],
            ],
            [
                'system_under_test' => new JwtClientTestHelper(
                    'example.com',
                    self::TOKEN,
                    'abc',
                    $this->getGuzzleConfig(ClientInterface::HTTP_METHOD_DELETE)
                ),
                'system_under_test_method' => 'httpDelete',
                'query_data' => [],
            ],
        ];
    }

    /**
     * Get the expected header keys that your client should have set on request.
     *
     * @return array
     */
    protected function getExpectedHeaderKeys()
    {
        $expectedHeaders = [
            'Authorization',
            'Content-Type',
            'x-api-key',
        ];

        return $expectedHeaders;
    }
}


tests/BasicClientTest.php
<?php

/**
 * Filename:    BasicClientTest.php
 * Created:     03/27/18, at 10:50 AM
 * @author      James Hollenbeck <jhollenbeck@nabancard.com>
 * @copyright   1992-2018 North American Bancard
 */

namespace NAB\Tests;

use NAB\Tests\ClientTestHelpers\BasicClientTestHelper;
use NAB\WebApiBaseSdk\ClientInterface;
use Psr\Http\Message\RequestInterface;

class BasicClientTest extends AbstractClientTestBase
{
    protected function getGuzzleConfig($expectedHttpMethod = '')
    {
        $testQueryParams = $this->GET_REQUEST_QUERY_PARAMS;
        $config = ['spy' => function (RequestInterface $request) use ($testQueryParams, $expectedHttpMethod) {
            $expectedHeaders = [
                'Authorization',
                'Content-Type',
                'x-api-key',
            ];
            $username = 'ima-user';
            $password = 'ima-password';
            $authHeaderDigest = base64_encode("{$username}:{$password}");

            $actualHeaders = array_keys($request->getHeaders());

            sort($actualHeaders);
            sort($expectedHeaders);
            $this->assertSame($expectedHeaders, $actualHeaders);

            $this->assertSame($expectedHttpMethod, $request->getMethod());

            $this->assertSame('abc', $request->getHeader('x-api-key')[0]);
            $this->assertSame('Basic ' . $authHeaderDigest, $request->getHeader('Authorization')[0]);

            if ($request->getMethod() === ClientInterface::HTTP_METHOD_GET) {
                $queryParams = $request->getUri()->getQuery();
                $this->assertSame(http_build_query($testQueryParams), $queryParams);
            }
        }];
        return $config;
    }

    public function clientCanHandleRequestDataProvider()
    {
        return [
            [
                'system_under_test' => new BasicClientTestHelper(
                    'example.com',
                    'ima-user',
                    'ima-password',
                    'abc',
                    $this->getGuzzleConfig(ClientInterface::HTTP_METHOD_GET)
                ),
                'system_under_test_method' => 'httpGet',
                'query_data' => $this->GET_REQUEST_QUERY_PARAMS,
            ],
            [
                'system_under_test' => new BasicClientTestHelper(
                    'example.com',
                    'ima-user',
                    'ima-password',
                    'abc',
                    $this->getGuzzleConfig(ClientInterface::HTTP_METHOD_POST)
                ),
                'system_under_test_method' => 'httpPost',
                'query_data' => [],
            ],
            [
                'system_under_test' => new BasicClientTestHelper(
                    'example.com',
                    'ima-user',
                    'ima-password',
                    'abc',
                    $this->getGuzzleConfig(ClientInterface::HTTP_METHOD_PUT)
                ),
                'system_under_test_method' => 'httpPut',
                'query_data' => []
            ],
            [
                'system_under_test' => new BasicClientTestHelper(
                    'example.com',
                    'ima-user',
                    'ima-password',
                    'abc',
                    $this->getGuzzleConfig(ClientInterface::HTTP_METHOD_DELETE)
                ),
                'system_under_test_method' => 'httpDelete',
                'query_data' => []
            ],
        ];
    }

    /**
     * Get the expected header keys that your client should have set on request.
     *
     * @return array
     */
    protected function getExpectedHeaderKeys()
    {
        $expectedHeaders = [
            'Authorization',
            'Content-Type',
            'x-api-key',
        ];

        return $expectedHeaders;
    }
}


tests/SimpleClientTest.php
<?php

/**
 * Filename:    SimpleClientTest.php
 * Created:     1/22/18, at 8:05 PM
 * @author      James Hollenbeck <jhollenbeck@nabancard.com>
 * @copyright   1992-2018 North American Bancard
 */

namespace NAB\Tests;

use NAB\Tests\ClientTestHelpers\SimpleClientTestHelper;
use NAB\WebApiBaseSdk\ClientInterface;

class SimpleClientTest extends AbstractClientTestBase
{
    public function clientCanHandleRequestDataProvider()
    {
        return [
            [
                'system_under_test' => new SimpleClientTestHelper(
                    'example.com',
                    'abc',
                    $this->getGuzzleConfig(ClientInterface::HTTP_METHOD_GET)
                ),
                'system_under_test_method' => 'httpGet',
                'query_data' => $this->GET_REQUEST_QUERY_PARAMS,
            ],
            [
                'system_under_test' => new SimpleClientTestHelper(
                    'example.com',
                    'abc',
                    $this->getGuzzleConfig(ClientInterface::HTTP_METHOD_POST)
                ),
                'system_under_test_method' => 'httpPost',
                'query_data' => [],
            ],
            [
                'system_under_test' => new SimpleClientTestHelper(
                    'example.com',
                    'abc',
                    $this->getGuzzleConfig(ClientInterface::HTTP_METHOD_PUT)
                ),
                'system_under_test_method' => 'httpPut',
                'query_data' => [],
            ],
            [
                'system_under_test' => new SimpleClientTestHelper(
                    'example.com',
                    'abc',
                    $this->getGuzzleConfig(ClientInterface::HTTP_METHOD_DELETE)
                ),
                'system_under_test_method' => 'httpDelete',
                'query_data' => [],
            ],
        ];
    }

    /**
     * Get the expected header keys that your client should have set on request.
     *
     * @return array
     */
    protected function getExpectedHeaderKeys()
    {
        $expectedHeaders = [
            'Content-Type',
            'x-api-key',
        ];

        return $expectedHeaders;
    }
}


lib/JwtClient.php
<?php
/**
 * Filename:    JwtClient.php
 * Created:     21/02/18, at 2:34 PM
 * @author      James Hollenbeck <jhollenbeck@nabancard.com>
 * @copyright   1992-2018 North American Bancard
 */

namespace NAB\WebApiBaseSdk;

use GuzzleHttp\Psr7\Request;
use NAB\WebApiBaseSdk\ClientInterface;
use NAB\WebApiBaseSdk\AbstractClient;

class JwtClient extends AbstractClient implements ClientInterface
{
    /**
     * @var string
     */
    private $jwt;

    /**
     * Sets up a new JWT Client.
     *
     * @param string      $baseUrl
     * @param string      $jwt
     * @param null|string $xApiKey
     * @param array       $config
     */
    public function __construct($baseUrl, $jwt, $xApiKey = null, array $config = [])
    {
        $this->jwt = $jwt;
        parent::__construct($baseUrl, $xApiKey, $config);
    }

    /**
     * Helper to create the final request object.
     *
     * @param string $method   The HTTP Method to use in the Request
     * @param string $endpoint The final endpoint (including the query params)
     * @param array  $headers  An array of headers to send
     * @param string $body     The body of the Request
     *
     * @return Request
     */
    protected function createRequest($method, $endpoint, array $headers, $body)
    {
        $headers['Authorization'] = 'Bearer ' . $this->jwt;
        $request = new Request($method, $endpoint, $headers, $body);

        return $request;
    }
}


lib/BasicClient.php
<?php
/**
 * Filename:    BasicClient.php
 * Created:     27/03/18, at 10:41 AM
 * @author      James Hollenbeck <jhollenbeck@nabancard.com>
 * @copyright   1992-2018 North American Bancard
 */

namespace NAB\WebApiBaseSdk;

use GuzzleHttp\Psr7\Request;
use NAB\WebApiBaseSdk\ClientInterface;
use NAB\WebApiBaseSdk\AbstractClient;

class BasicClient extends AbstractClient implements ClientInterface
{
    /**
     * @var string
     */
    private $username;

    /**
     * @var string
     */
    private $password;

    /**
     * Sets up a new Basic Client.
     *
     * @param string      $baseUrl
     * @param string      $username
     * @param string      $password
     * @param null|string $xApiKey
     * @param array       $config
     */
    public function __construct($baseUrl, $username, $password, $xApiKey = null, array $config = [])
    {
        $this->username = $username;
        $this->password = $password;
        parent::__construct($baseUrl, $xApiKey, $config);
    }

    /**
     * Helper to create the final request object.
     *
     * @param string $method   The HTTP Method to use in the Request
     * @param string $endpoint The final endpoint (including the query params)
     * @param array  $headers  An array of headers to send
     * @param string $body     The body of the Request
     *
     * @return Request
     */
    protected function createRequest($method, $endpoint, array $headers, $body)
    {
        $headers['Authorization'] = 'Basic ' . base64_encode("{$this->username}:{$this->password}");
        $request = new Request($method, $endpoint, $headers, $body);

        return $request;
    }
}


lib/SignatureV4.php
<?php

/**
 * Filename:    SignatureV4.php
 * Created:     1/22/18, at 8:39 PM
 * @author      James Hollenbeck <jhollenbeck@nabancard.com>
 * @copyright   1992-2018 North American Bancard
 */

namespace NAB\WebApiBaseSdk;

use Aws\Signature\SignatureV4 as AwsSignatureV4;

/**
 * An extension of Amazon's SignatureV4 class, that allows us to have sensible defaults
 * we are likely to use most of the time in-house.
 */
class SignatureV4 extends AwsSignatureV4
{
    /**
     * Sets up a new SignatureV4 with sensible defaults.
     *
     * @param string $service Service name to use when signing
     * @param string $region  Region name to use when signing
     * @param array $options  Array of configuration options used when signing
     *                          - unsigned-body: Flag to make request have unsigned payload.
     *                            Unsigned body is used primarily for streaming requests.
     */
    public function __construct($service = 'execute-api', $region = 'us-east-1', array $options = [])
    {
        parent::__construct($service, $region, $options);
    }
}


lib/SimpleClient.php
<?php
/**
 * Filename:    SimpleClient.php
 * Created:     21/02/18, at 6:34 PM
 * @author      James Hollenbeck <jhollenbeck@nabancard.com>
 * @copyright   1992-2018 North American Bancard
 */

namespace NAB\WebApiBaseSdk;

use GuzzleHttp\Psr7\Request;
use NAB\WebApiBaseSdk\AbstractClient;

/**
 * Simple client sans authentication standards (with exception of possible x-api-key, if you so wish).
 */
class SimpleClient extends AbstractClient
{
    /**
     * Helper to create the final request object.
     *
     * @param string $method   The HTTP Method to use in the Request
     * @param string $endpoint The final endpoint (including the query params)
     * @param array  $headers  An array of headers to send
     * @param string $body     The body of the Request
     *
     * @return Request
     */
    public function createRequest($method, $endpoint, array $headers, $body)
    {
        return new Request($method, $endpoint, $headers, $body);
    }
}


lib/ClientInterface.php
<?php
/**
 * Filename:    ClientInterface.php
 * Created:     21/02/18, at 2:43 PM
 * @author      James Hollenbeck <jhollenbeck@nabancard.com>
 * @copyright   1992-2018 North American Bancard
 */

namespace NAB\WebApiBaseSdk;

/**
 * While it's certainly not the normal case you see interfaces acting as an enum - in this case, since HTTP
 * verbs are unlikely to grow to unmanageable bounds, due to them being relatively static for so many years,
 * we can get away with this here. Otherwise - if you need many different states/statuses, look into strategy or
 * perhaps flyweight patterns to maintain value-object states.
 */
interface ClientInterface
{
    const HTTP_METHOD_POST = 'POST';
    const HTTP_METHOD_GET = 'GET';
    const HTTP_METHOD_PUT = 'PUT';
    const HTTP_METHOD_DELETE = 'DELETE';

    /**
     * Wrapper around Guzzle's sending process of a get request.
     *
     * @param string $path       The path to send the request to (no query params)
     * @param array $queryParams Any query params to attach to the url
     * @param array $headers     Extra headers to add to Guzzle's default
     * @param mixed $body        The body of the get request
     *
     * @return mixed|\Psr\Http\Message\ResponseInterface
     */
    public function httpGet($path, array $queryParams = [], array $headers = [], $body = null);

    /**
     * Wrapper around Guzzle's sending process of a post request.
     *
     * @param string $path       The path to send the request to (no query params)
     * @param array $queryParams Any query params to attach to the url
     * @param array $headers     Extra headers to add to Guzzle's default
     * @param mixed $body        The body of the post request
     *
     * @return mixed|\Psr\Http\Message\ResponseInterface
     */
    public function httpPost($path, array $queryParams = [], array $headers = [], $body = null);

    /**
     * Wrapper around Guzzle's sending process of a put request.
     *
     * @param string $path       The path to send the request to (no query params)
     * @param array $queryParams Any query params to attach to the url
     * @param array $headers     Extra headers to add to Guzzle's default
     * @param mixed $body        The body of the put request
     *
     * @return mixed|\Psr\Http\Message\ResponseInterface
     */
    public function httpPut($path, array $queryParams = [], array $headers = [], $body = null);

    /**
     * Wrapper around Guzzle's sending process of a delete request.
     *
     * @param string $path       The path to send the request to (no query params)
     * @param array $queryParams Any query params to attach to the url
     * @param array $headers     Extra headers to add to Guzzle's default
     * @param mixed $body        The body of the delete request
     *
     * @return mixed|\Psr\Http\Message\ResponseInterface
     */
    public function httpDelete($path, array $queryParams = [], array $headers = [], $body = null);
}


lib/AwsApiGatewayClient.php
<?php

/**
 * Filename:    AwsApiGatewayClient.php
 * Created:     1/22/18, at 7:55 PM
 * @author      James Hollenbeck <jhollenbeck@nabancard.com>
 * @copyright   1992-2018 North American Bancard
 */

namespace NAB\WebApiBaseSdk;

use Aws\Credentials\Credentials;
use NAB\WebApiBaseSdk\SignatureV4;
use NAB\WebApiBaseSdk\AbstractClient;
use GuzzleHttp\Psr7\Request;

/**
 * Client which will allow us to sign requests we make using sigv4.
 */
class AwsApiGatewayClient extends AbstractClient
{
    /**
     * @var Credentials
     */
    private $credentials;

    /**
     * @var SignatureV4
     */
    private $signature;

    /**
     * Sets up a new Client.
     *
     * @param string      $baseUrl     The base-url for the request, no trailing '/'
     * @param SignatureV4 $signature   An instance of AWS' SigV4 signer
     * @param Credentials $credentials The access key and secret key for an IAM user
     * @param string      $xApiKey     An API key to use during requests, if needed
     * @param array $config            Guzzle configuration
     */
    public function __construct(
        $baseUrl,
        SignatureV4 $signature,
        Credentials $credentials,
        $xApiKey = null,
        array $config = []
    ) {
        $this->signature = $signature;
        $this->credentials = $credentials;

        parent::__construct($baseUrl, $xApiKey, $config);
    }

    /**
     * Helper to create the final request object.
     *
     * @param string $method   The HTTP Method to use in the Request
     * @param string $endpoint The final endpoint (including the query params)
     * @param array  $headers  An array of headers to send
     * @param string $body     The body of the Request
     *
     * @return Request
     */
    protected function createRequest($method, $endpoint, array $headers, $body)
    {
        $request = new Request($method, $endpoint, $headers, $body);
        $signedRequest = $this->signature->signRequest($request, $this->credentials);

        return $signedRequest;
    }
}


lib/AbstractClient.php
<?php
/**
 * Filename:    AbstractClient.php
 * Created:     21/02/18, at 2:55 PM
 * @author      James Hollenbeck <jhollenbeck@nabancard.com>
 * @copyright   1992-2018 North American Bancard
 */

namespace NAB\WebApiBaseSdk;

use GuzzleHttp\Client;
use GuzzleHttp\Psr7\Request;
use GuzzleHttp\Psr7\Response;
use NAB\WebApiBaseSdk\ClientInterface;

/**
 * Parent of any client this library supports regarding authentication.
 */
abstract class AbstractClient extends Client implements ClientInterface
{
    /**
     * @var string
     */
    private $baseUrl;

    /**
     * @var array
     */
    private $headers;

    /**
     * Sets up a new Client implementation. We are being restrictive here for aid in debugging and control
     * over guzzle changes as they change implementations.
     *
     * @param string $baseUrl The base URL, without trailing / or query params.
     * @param string $xApiKey An API key header (usually associated with AWS APIG)
     * @param array  $config  Any Guzzle configuration
     */
    public function __construct($baseUrl, $xApiKey = null, array $config = [])
    {
        $this->baseUrl = $baseUrl;

        // By default, let's not throw exceptions on non-200 responses.
        if (!isset($config['http_errors'])) {
            $config['http_errors'] = false;
        }

        $headers = ['Content-Type' => 'application/json'];
        if (isset($config['headers'])) {
            $headers += $config['headers'];
            unset($config['headers']);
        }

        if ($xApiKey) {
            $headers['x-api-key'] = $xApiKey;
        }

        $this->headers = $headers;
        parent::__construct($config);
    }

    /**
     * A special little gift for this base SDK. If you ever come across the need where your `one` SDK is actually
     * the same `path` spread across more than one host - you can swap hosts on run-time to allow you to aggregate
     * results from the same SDK rather than creating new services over and over.
     *
     * NOTE: Because this class will most-likely be wrapped up by your actual SDK - you are going to have to
     * have a function, such as `getClient`, that returns the current instance.
     *
     * @param string $newBaseUrl The new base-url - no trailing `/`.
     */
    public function swapBaseUrl($newBaseUrl)
    {
        $this->baseUrl = $newBaseUrl;
    }

    /**
     * Helper to create the final request object.
     *
     * @param string $method   The HTTP Method to use in the Request
     * @param string $endpoint The final endpoint (including the query params)
     * @param array  $headers  An array of headers to send
     * @param string $body     The body of the Request
     *
     * @return Request
     */
    abstract protected function createRequest($method, $endpoint, array $headers, $body);

    /**
     * Wrapper around Guzzle's sending process of a post request.
     *
     * @param string $path       The path to send the request to (no query params)
     * @param array $queryParams Any query params to attach to the url
     * @param array $headers     Extra headers to add to Guzzle's default
     * @param mixed $body        The body of the post request
     *
     * @return mixed|\Psr\Http\Message\ResponseInterface
     *
     * @throws \GuzzleHttp\Exception\GuzzleException IFF HTTP_ERRORS are true on construct.
     */
    public function httpPost($path, array $queryParams = [], array $headers = [], $body = null)
    {
        return $this->doSendRequest(self::HTTP_METHOD_POST, $path, $queryParams, $headers, $body);
    }

    /**
     * Wrapper around Guzzle's sending process of a get request.
     *
     * @param string $path       The path to send the request to (no query params)
     * @param array $queryParams Any query params to attach to the url
     * @param array $headers     Extra headers to add to Guzzle's default
     * @param mixed $body        The body of the get request
     *
     * @return mixed|\Psr\Http\Message\ResponseInterface
     *
     * @throws \GuzzleHttp\Exception\GuzzleException IFF HTTP_ERRORS are true on construct.
     */
    public function httpGet($path, array $queryParams = [], array $headers = [], $body = null)
    {
        return $this->doSendRequest(self::HTTP_METHOD_GET, $path, $queryParams, $headers, $body);
    }

    /**
     * Wrapper around Guzzle's sending process of a put request.
     *
     * @param string $path       The path to send the request to (no query params)
     * @param array $queryParams Any query params to attach to the url
     * @param array $headers     Extra headers to add to Guzzle's default
     * @param mixed $body        The body of the put request
     *
     * @return mixed|\Psr\Http\Message\ResponseInterface
     *
     * @throws \GuzzleHttp\Exception\GuzzleException IFF HTTP_ERRORS are true on construct.
     */
    public function httpPut($path, array $queryParams = [], array $headers = [], $body = null)
    {
        return $this->doSendRequest(self::HTTP_METHOD_PUT, $path, $queryParams, $headers, $body);
    }

    /**
     * Wrapper around Guzzle's sending process of a delete request.
     *
     * @param string $path       The path to send the request to (no query params)
     * @param array $queryParams Any query params to attach to the url
     * @param array $headers     Extra headers to add to Guzzle's default
     * @param mixed $body        The body of the delete request
     *
     * @return mixed|\Psr\Http\Message\ResponseInterface
     *
     * @throws \GuzzleHttp\Exception\GuzzleException IFF HTTP_ERRORS are true on construct.
     */
    public function httpDelete($path, array $queryParams = [], array $headers = [], $body = null)
    {
        return $this->doSendRequest(self::HTTP_METHOD_DELETE, $path, $queryParams, $headers, $body);
    }

    /**
     * @return string The base url (no trailing /).
     */
    protected function getBaseUrl()
    {
        return $this->baseUrl;
    }

    /**
     * @return array Any set headers.
     */
    protected function getHeaders()
    {
        return $this->headers;
    }

    /**
     * Wrapper around Guzzle's sending process of a request.
     *
     * @param string $method     The HTTP method of the request
     * @param string $path       The path to send the request to (no query params)
     * @param array $queryParams Any query params to attach to the url
     * @param array $headers     Extra headers to add to Guzzle's default
     * @param mixed $body        The body of the post request
     *
     * @return mixed|\Psr\Http\Message\ResponseInterface
     *
     * @throws \GuzzleHttp\Exception\GuzzleException IFF HTTP_ERRORS are true on construct.
     */
    private function doSendRequest($method, $path, array $queryParams, array $headers, $body)
    {
        if ($queryParams) {
            $path = $path . '?' . http_build_query($queryParams);
        }

        $endpoint = $this->getBaseUrl() . $path;
        $requestHeaders = array_merge($this->getHeaders(), $headers);
        $request = $this->createRequest($method, $endpoint, $requestHeaders, $body);

        $response = $this->send($request);
        return $response;
    }
}
